services/vector_db_service.py

```
@@ -47,6 +47,27 @@ def upsert_vectors(self, vectors: List[List[float]], ids: List[str]):
         if not self.index:
             raise RuntimeError("Index is not initialized. Call create_index first.")
         self.index.upsert(vectors=[(id, vec) for id, vec in zip(ids, vectors)])
+    
+    def upsert(self, namespace, ids, vectors, metadata=None):
+        """
+        Upsert vectors into the index, ensuring index is created and metadata is validated.
+        """
+        if self.index is None:
+            self.create_index()
+        if not (len(ids) == len(vectors)):
+            raise ValueError("ids and vectors must have the same length")
```

Ruff (0.13.1)

Avoid specifying long messages outside the exception class

(TRY003)